### PR TITLE
Add support for SFTP

### PR DESCRIPTION
**Pull request by martinmaillard**
Add a `protocol` option to each command that can be either `ftp` (default) or `sftp`.

I would still like to test this with a "real" SFTP server because I don't trust the implementation of the test server I used.

Note that it currently only supports user/password authentication. Not sure what we need to support.

### CODE REVIEW

**Comment by martinmaillard** [comment]
I tried `list` with a real SFTP server and it works.

File: lib/SftpClient.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      port: this.port,
+      username: this.user,
+      password: this.password
+    })
+
+    return this.session
+  }
+
+  async disconnect () {
+    if (this.session !== null) {
```
Comment (bergos) [inline lib/SftpClient.js]: It should be replaced by `if (this.session) {`. That's a more common pattern in JS to also catch `undefined`, uninitialized variables.

File: lib/SftpClient.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+const SFTP = require('sftp-promises')
+const FileParser = require('ftp/lib/parser')
+
+class SftpClient {
+  constructor ({ host, port = 21, user, password }) {
+    this.host = host
+    this.port = port
+    this.user = user
+    this.password = password
+    this.client = new SFTP()
```
Comment (bergos) [inline lib/SftpClient.js:11]: Also add `this.session = null` at the end. Initializing all member variables in the constructor gives a better overview of all possible variables, makes debugging easier and also helps the VM optimizing.

File: test/list.test.js [hunk 3]
```diff
@@ -2,34 +2,27 @@
...
-    const server = new FtpServer({ user: 'test', password: '1234' })
-    await server.start()
-
-    const stream = await list({ pathname: 'data', ...server.options })
-    const filenames = await getStream.array(stream)
-
-    await server.stop()
-
-    expect(filenames).toEqual(['data/abc.txt', 'data/xyz.txt'])
+  each([
```
Comment (bergos) [inline test/list.test.js]: Good idea to remove redundancy, but I don't think the `jest-each` package is required and the labels can be more readable:

```js
[{
  label: 'on a ftp server with anonymous user',
  factory: () => new FtpServer()
}].forEach(({ label, factory } => {
  it(`lists files from the given directory ${label}`, () => {
    // ...
  })
})
```

File: test/support/server.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+
+async function withServer (serverFactory, callback) {
```
Comment (bergos) [inline test/support/server.js:2]: Good idea! 👍

**Review by bergos — CHANGES_REQUESTED** [review]

**Comment by martinmaillard** [comment]
Thanks for the feedback @bergos, I think I addressed all your comments.

**Comment by martinmaillard** [comment]
So `read` fails on Linux.

The interesting thing is that my implementation fails (timeout):
```javascript
async read (path) {
  return this.client.createReadStream(path, this.session)
}
```
but if I replace it with this it works:
```javascript
async read (path) {
  const buffer = await this.client.getBuffer(path, this.session)
  const stream = new Readable()
  stream.push(buffer)
  stream.push(null)
  return stream
}
```

There seems to be an issue with `createReadStream`. I have to check the implementation in "ssh2" to find out why.

**Comment by martinmaillard** [comment]
I tried two things to isolate the issue:

Used `sftp.getStream` instead of `sftp.createReadStream`. It does solve the issue.
```javascript
async read (path) {
  const stream = new MyDuplex()
  await this.client.getStream(path, stream, this.session)
  return stream
}
```

Re-implemented the server with `node-sftp-server` instead of `sftp-fs`. Got the same timeout issue.

I think the problem is not on the server, but on the client. It has to do with `sftp.createReadStream`, which uses the same function from the "ssh2-streams" lib. Not sure what to do next...

**Comment by martinmaillard** [comment]
Replaced `createReadStream` with `getStream` and a custom Duplex stream to flow data. It should work now but it would be great if you could take a look at my `Pipe` stream implementation @bergos

File: lib/SftpClient.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    })
+  }
+
+  async move (source, target) {
+    return this.client.mv(source, target, this.session)
+  }
+
+  async read (path) {
+    // Using `getStream` because `createReadStream` causes issues on Linux
+    const stream = new Pipe()
```
Comment (bergos) [inline lib/SftpClient.js]: Did you try [PassThrough](https://nodejs.org/api/stream.html#stream_class_stream_passthrough)? It's also included in the `readable-stream` package. If for some reason it doesn't work, add a comment why and put the `Pipe` class into a separate file. When that's done, the PR can be merged.

**Review by bergos — APPROVED** [review]

File: lib/SftpClient.js (hunk 5 above)
Comment (martinmaillard) [reply lib/SftpClient.js]: I missed PassThrough when reading the docs. It looks like exactly what I was looking for, nice!
I'll change it and merge.

**Pull request merged**